doc/sections/subsections/parameters.rst
.. _params_yml:

Simulation parameters
---------------------

All Struphy simulation parameters can be set in the ``.yml`` parameter file. For each model,
a default file can be created via::

    struphy params MODEL

The file is saved in the current input path (``struphy -p``, under ``/inp``) 
and can be modified and renamed. An example of such a file is given 
`here <https://gitlab.mpcdf.mpg.de/struphy/struphy/-/blob/devel/src/struphy/io/inp/parameters.yml?ref_type=heads>`_.

Struphy simulation parameters fall into three classes, addressed by
up to 8 different keys in the ``.yml`` file:

1. Discretization parameters:

* :ref:`grid` (under the key ``grid``)
* :ref:`time` (under the key ``time``)

2. Physics parameters:

* :ref:`units` (under the key ``units``)
* :ref:`geometry` (under the key ``geometry``)
* :ref:`fluid_background` (under the key ``fluid_background``)

3. Species parameters:

* :ref:`em_fields` (under the key ``em_fields``)
* :ref:`fluid` (under the key ``fluid``)
* :ref:`kinetic` (under the key ``kinetic``)

The ``.yml`` parameter file has a generic structure for all Struphy models.
The keys ``grid``, ``time``, ``units`` and ``geometry`` are present 
in every model; moreover, at least one of the three
species types ``em_fields``, ``fluid`` and ``kinetic`` must be present 
in each model, relating to the model variable(s). An ``fluid_background``
is not mandatory.

The structure of the dictionaries under each of the 8 top-level keys
is discussed below. Special information is available on how to set :ref:`initial_conditions`
and :ref:`boundary_conditions`. 

Some hints for editing a parameter file:

* Strings can either be set as e.g. ``'Cuboid'`` or ``Cuboid``, i.e. with or without quotes - both works.
* The parameter ``null`` will be transformed to Python's ``None`` type.
* Available geometries can be found in :ref:`avail_mappings`
* Available fluid equilibria can be found in :ref:`equils`
* Available kinetic backgrounds can be found in :ref:`kinetic_backgrounds`
* Available fluid backgrounds can be found in :ref:`equils_avail`
* Available perturbations can be found in :ref:`avail_inits`

.. _grid:

Space grid parameters
^^^^^^^^^^^^^^^^^^^^^

Struphy uses a tensor-product grid in three dimensions for space discretization.

Example:

::

    grid :
        Nel          : [12, 14, 4] 
        p            : [3, 4, 2]  
        spl_kind     : [False, True, True] 
        dirichlet_bc : [[False, False], [False, False], [False, False]] 
        dims_mask    : [True, True, True] 
        nq_el        : [2, 2, 2] 
        nq_pr        : [2, 2, 2] 
        polar_ck     : -1 

Parameters:

.. list-table:: 
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``Nel``
     - Number of elements (grid cells) in each direction.
     - 3-list[int]
     - 
   * - ``p``
     - Spline degree in each direction.
     - 3-list[int]
     -
   * - ``spl_kind``
     - Kind of spline in each direction (see :ref:`uni_variate_spaces`).
     - 3-list[bool]
     - * ``True``: periodic
       * ``False``: clamped
   * - ``dirichlet_bc``
     - Homogeneous Dirichlet boundary conditions at left and/or right boundary in each direction.
     - 3-list[2-list[int]]
     - * ``True``: yes
       * ``False``: no
       * See :ref:`feec_bcs` for how to set boundary conditions.
   * - ``dims_mask``
     - Allow domain decomposition of direction.
     - 3-ist[bool]
     - * ``True``: yes
       * ``False``: no
   * - ``nq_el``
     - Number of quadrature points per element (e.g. for L2-projections) in each direction.
     - List[int]
     -
   * - ``nq_pr``
     - Quadrature points between Greville points (for commuting projectors) in each direction. 
     - List[int]
     -
   * - ``polar_ck``
     - :math:`C^k` smoothness of B-splines at polar singularity :math:`\eta_1=0` for polar geometries.
     - int
     - * ``-1``: standard tensor product, not continuous at pole
       * ``1``: use :math:`C^1` :ref:`polar_splines`

These paramters are primarily used to instantiate :class:`struphy.feec.psydac_derham.Derham`.


.. _time:

Time stepping parameters
^^^^^^^^^^^^^^^^^^^^^^^^

Example:

::

    time :
        dt         : 0.005 
        Tend       : 0.015 
        split_algo : LieTrotter 

Parameters:

.. list-table:: 
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``dt``
     - Time step in units given by a model's :ref:`normalization` (inspect via ``struphy units -i <params_file> MODEL`` from the console). 
     - float
     -
   * - ``Tend``
     - End time in same units as ``dt``.
     - float
     -
   * - ``split_algo``
     - Time splitting algorithm
     - str
     - * ``LieTrotter``: first order
       * ``Strang``: second order, symmetric


.. _units:

Units
^^^^^

Example:

::

    units : 
        x : 1.  
        B : 1. 
        n : 1. 
        kBT : 1. 

Parameters:

.. list-table:: 
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``x``
     - Length scale unit in Meter (m).
     - float
     -
   * - ``B``
     - Magnetic field unit in Tesla (T).
     - float
     -
   * - ``n``
     - Number density unit in :math:`10^{20}\ m^{-3}`.
     - float
     -
   * - ``kBT`` (optional)
     - Thermal energy unit in keV (optional).
     - float
     -

Theses base units are used to derive all other units via :func:`struphy.io.setup.derive_units`.
See also :ref:`normalization` for details.


.. _geometry:

Geometry
^^^^^^^^

Example:

::

    geometry :
        type : Cuboid 
        Cuboid : {}

Parameters:

.. list-table:: 
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``type``
     - The mapping :math:`F:[0, 1]^3 \to \Omega` (its class name ``<mapping_class>``). 
     - str
     - See :ref:`avail_mappings`.
   * - ``<mapping_class>``
     - Parameters for the chosen mapping. If empty, the defaults are taken.
     - dict
     - See docstrings in :ref:`avail_mappings`.


.. _fluid_background:

Fluid background
^^^^^^^^^^^^^^^^

Example:

::

    fluid_background :
        HomogenSlab : {}

Parameters:

.. list-table:: 
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<equil_class>``
     - Class name and parameters for the equilibrium. If empty, the defaults are taken.
     - dict
     - See :ref:`equils_avail`.


.. _em_fields:

Electromagnetic fields
^^^^^^^^^^^^^^^^^^^^^^

Initial conditions are the sum of ``background`` + ``perturbation``, see :ref:`initial_conditions`.

Example:

::

    em_fields :
        background: 
            phi : 
              LogicalConst :
                  values : 1.3 
            A : 
              LogicalConst :
                  values : [.3, .15, null] 
        perturbation :
            phi:
              TorusModesCos :
                  given_in_basis : '0' 
                    A : 
                  ms : [1] 
            A:
              TorusModesCos :
                  given_in_basis : [null, 'v', null]
                  ms : [null, [1, 3], null]  

Parameters:

.. list-table:: ``background``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<variable_names>``
     - Name of variable to be initialized.
     - dict
     - Contains class name and parameters for the static background, see :ref:`equils_avail`. If empty, the defaults are taken.

.. list-table:: ``perturbation``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<variable_names>``
     - Name of variable to be initialized.
     - dict
     - Contains class name and parameters for the perturbation, see :ref:`avail_inits`. If empty, the defaults are taken.



.. _fluid:

Fluid variables
^^^^^^^^^^^^^^^

There can be multiple ``fluid`` species in a Struphy model,
each of which has its paramaters under a key ``<species_name>``.
For each species, there are

* physical parameters ``phys_params``,
* initial conditions as the sum of ``background`` + ``perturbation``, see :ref:`initial_conditions`.

Example (one fluid species):

::

    fluid :
        <species_name> :
            phys_params:
                A : 1 
                Z : 1 
            background : 
                velocity :
                    LogicalConst :
                        values : [null, 1.5, null] 
                density : 
                    LogicalConst :
                        values : 2.3 
            perturbation :
                density : 
                    TorusModesCos :
                        given_in_basis : '0' 
                        ms : [1, 3] 

Parameters:

.. list-table:: ``phys_params``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``A``
     - Mass number in units of proton mass.
     - float
     - :math:`A=1/1836` for electrons.
   * - ``Z``
     - Signed charge number in units of elementary charge.
     - int
     - :math:`Z=-1` for electrons.

.. list-table:: ``background``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<variable_names>``
     - Name of variable to be initialized.
     - dict
     - Contains class name and parameters for the static background, see :ref:`equils_avail`. If empty, the defaults are taken.

.. list-table:: ``perturbation``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<variable_names>``
     - Name of variable to be initialized.
     - dict
     - Contains class name and parameters for the perturbation, see :ref:`avail_inits`. If empty, the defaults are taken.

.. _kinetic:

Kinetic variables
^^^^^^^^^^^^^^^^^


There can be multiple ``kinetic`` species in a Struphy model,
each of which has its paramaters under a key ``<species_name>``.
For each species, there are

* physical parameters ``phys_params``,
* information on the ``markers``,
* parameters for saving kinetic data ``save_data``,
* initial conditions as the sum of ``background`` + ``perturbation``, see :ref:`initial_conditions`.

For ``kinetic`` species, the ``background`` is mandatory.

Example (one kinetic species):

::

    kinetic :
        <species_name> :
            phys_params :
                A : 1
                Z : 1
            markers :
                type    : full_f 
                ppc     : null 
                Np      : 1000
                bufsize     : .25
                bc : 
                    type    : [remove, periodic, periodic] 
                    remove  : boundary_transfer
                loading :
                    type              : pseudo_random
                    seed              : 1234 
                    moments           : [0., 0., 0., 1., 1., 1.] 
                    spatial           : uniform
                    dir_external      : 'path_to_particles' 
                    dir_particles     : 'path_to_particles' 
                    dir_particles_abs : 'path_to_particles' 
            save_data :
                n_markers : 3
                f :
                    slices : [v1, e1_v1]
                    n_bins : [[32], [32, 32]]
                    ranges : [[[-3., 3.]], [[0., 1.], [-5., 5.]]]
            background : 
                Maxwellian3D :
                    n  : 0.05
                    u2 : 2.5
            perturbation :
                n :
                    TorusModesCos :
                        given_in_basis :  '0' 
                        ms : [1, 3] 

Parameters:

.. list-table:: ``phys_params``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``A``
     - Mass number in units of proton mass.
     - float
     - :math:`A=1/1836` for electrons.
   * - ``Z``
     - Signed charge number in units of elementary charge.
     - int
     - :math:`Z=-1` for electrons.

.. list-table:: ``markers``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``Np``
     - Total number of markers.
     - int
     - 
   * - ``ppc``
     - Number of markers per grid cell.
     - int
     - Takes effect only if ``Np`` is absent.
   * - ``ppb``
     - Number of markers per sorting box.
     - int
     - Takes effect only if both ``Np`` and ``ppc`` are absent.
   * - ``bufsize``
     - Size of MPI-buffer in markers array as fraction of markers per process.
     - float
     - 
   * - ``bc``
     - Marker boundary conditions. See :meth:`struphy.pic.base.Particles.apply_kinetic_bc`.
     - dict
     - * ``type`` (3-list[str]): marker boundary conditions in each logical direction.
     
         * ``remove``: markers outside of :math:`\Omega` are removed or re-filled.
         * ``reflect``: markers are reflected at :math:`\partial \Omega`.
         * ``periodic``: markers re-enter at other side.
       * ``remove`` (str): re-fills markers in radial direction on same flux surface if ``type`` is ``remove``.

         * ``boundary_transfer``:  when particles reach the INNER radial boundary, transfer them to the opposite poloidal angle of the same magnetic flux surface.
         * ``particle_refilling``: when particles reach the OUTER radial boundary, transfer them to the opposite poloidal angle of the same magnetic flux surface.
   * - ``loading``
     - How to load random markers.
     - dict
     - * ``type`` (str): particle loading mechanism, see :meth:`struphy.pic.base.Particles.draw_markers`.

         * ``pseudo_random`` load markers with standard random generator ``np.random.rand``.
         * ``sobol_standard`` load Sobol sequence, see :module:`struphy.pic.sobol_seq`.
         * ``sobol_antithetic`` load Sobol sequence, see :module:`struphy.pic.sobol_seq`.
         * ``external``: load markers from external .hdf5 file specified in ``dir_external``.
         * ``restart``: load markers from stopped simulation in folder ``dir_particles`` or ``dir_particles_abs``.
       * ``seed`` (int): seed for random number generator ``np.random.rand``.
       * ``moments`` (list): mean velocity and standard deviation of Gaussian that markers are drawn from, see :meth:`struphy.pic.base.Particles.draw_markers`.
       * ``spatial`` (str): uniformity of markers in position space.
         
         * ``uniform``: markers are uniform on the logical unit cube.
         * ``disc``: markers are uniform on the disc :math:`(\eta_1, \eta_2) \mapsto (r, \theta)`.
       * optional: ``dir_external`` (str): path to .hdf5 file
       * optional: ``dir_particles`` (str): simulation folder relative to current output path 
       * optional: ``dir_particles_abs`` (str): simulation folder, absolute path
   * - ``control_variate``
     - Whether to use a :ref:`control_var` for noise reduction (only if ``type`` is ``full_f`` or ``sph``).
     - bool
     -


.. list-table:: ``save_data``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``n_markers``
     - Number of markers to be saved during simulation. Markers with an ID less than ``n_markers`` are saved.
     - int
     - 
   * - ``f``
     - Binning plots of the distribution function in phase space.
     - dict
     - * ``slices`` (list): in which directions to bin (e.g. ``[e1_e2, e1_v2_v3]``)
       * ``n_bins`` (list): number of bins in each slice-direction (e.g. ``[[16, 20], [16, 18, 22]]``)
       * ``ranges`` (list): bin range in each slice-direction (e.g. [[[0., 1.], [0., 1.]], [[0, 1.], [-4., 4.], [-5., 5.]]])


.. list-table:: ``background``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<bckgr_class>``
     - Class name and parameters for the static kinetic background. If empty, the defaults are taken.
     - dict
     - See :ref:`kinetic_backgrounds`.

.. list-table:: ``perturbation``
   :widths: 15 35 10 40
   :header-rows: 1

   * - Name
     - Description
     - Format
     - Choices
   * - ``<moment_names>``
     - Name of moment to be initialized.
     - dict
     - Contains class name and parameters for the perturbation, see :ref:`avail_inits`. If empty, the defaults are taken.

CHANGELOG.md
## Version 2.4.5

### Headlines

* Enable `pyccel 2.0` (https://github.com/pyccel/pyccel) !678
* New interface to `gvec` Python API (https://gitlab.mpcdf.mpg.de/gvec-group/gvec) !705 and !723

### Other user news

* Rename `Particles` parameter `eps -> bufsize` for more clarity !690
* Particle simulations: save weights together with orbits !690
* Add `mpi_dims_mask` for mesh-less simulations !690
* Added `gravity` and `thermodynamics` (isothermal or polytropic) options to `PushVinSPHpressure` !690
* Add options `gaussian_xy` and `step_function_x` to the `ConstanVelocity` fluid background !690
* Allow for running `struphy run` without specifying the model if `model: MODEL` is specified in the params !708
* Remove the `-d` flag from `struphy pproc`, now just run `struphy pproc sim_1 sim_2 ...` !709
* New variational MHD model with $q = \sqrt{p}$ as a variable !697

### Developer news

* New methods in `Particles`: `n_mks_on_each_proc`, `n_mks_on_clone`, `n_mks_on_each_clone` and `n_mks_global`; they refer to valid markers (not holes or ghosts) !690
* The CI pipeline no longer runs on every push, only on pushes to branches which currently have a MR with devel or master !719

### Bug fixes

* Print user warning and temrinate MPI when out-of-bounds error occurs due to load imbalance in particles comm !690
* Do NOT duplicate marker IDs when working with clones !690
* Enable running `cprofile` without a grid !690


src/struphy/io/options.py
from dataclasses import dataclass
from typing import Literal, get_args
import numpy as np
import os

from struphy.physics.physics import ConstantsOfNature


## Literal options

# time
SplitAlgos = Literal["LieTrotter", "Strang"]

# derham
PolarRegularity = Literal[-1, 1]
OptsFEECSpace = Literal["H1", "Hcurl", "Hdiv", "L2", "H1vec"]
OptsVecSpace = Literal["Hcurl", "Hdiv", "H1vec"]

# fields background
BackgroundTypes = Literal["LogicalConst", "FluidEquilibrium"]

# perturbations
NoiseDirections = Literal["e1", "e2", "e3", "e1e2", "e1e3", "e2e3", "e1e2e3"]
GivenInBasis = Literal['0', '1', '2', '3', 'v', 'physical', 'physical_at_eta', 'norm', None]

# solvers
OptsSymmSolver = Literal["pcg", "cg"]
OptsGenSolver = Literal["pbicgstab", "bicgstab"]
OptsMassPrecond = Literal["MassMatrixPreconditioner", None]

# markers
OptsPICSpace = Literal["Particles6D", "DeltaFParticles6D", "Particles5D", "Particles3D"]
OptsMarkerBC = Literal["periodic", "reflect"]
OptsRecontructBC = Literal["periodic", "mirror", "fixed"]
OptsLoading = Literal["pseudo_random", 'sobol_standard', 'sobol_antithetic', 'external', 'restart', "tesselation"]
OptsSpatialLoading = Literal["uniform", "disc"]
OptsMPIsort = Literal["each", "last", None]


## Option classes

@dataclass
class Time:
    """Time stepping options.

    Parameters
    ----------
    dt : float
        Time step.
        
    Tend : float
        End time.
        
    split_algo : SplitAlgos
        Splitting algorithm (the order of the propagators is defined in the model).
    """

    dt: float = 0.01
    Tend: float = 0.03
    split_algo: SplitAlgos = "LieTrotter"

    def __post_init__(self):
        check_option(self.split_algo, SplitAlgos)


@dataclass
class BaseUnits:
    """
    Base units are passed to __init__, other units derive from these.

    Parameters
    ----------
    x : float
        Unit of length in meters.

    B : float
        Unit of magnetic field in Tesla.

    n : float
        Unit of particle number density in 1e20/m^3.

    kBT : float, optional
        Unit of internal energy in keV. 
        Only in effect if the velocity scale is set to 'thermal'.
    """
    x: float = 1.0
    B: float = 1.0
    n: float = 1.0
    kBT: float = None


class Units:
    """
    Colllects base units and derives other units from these.
    """

    def __init__(self, base: BaseUnits = None):
        if base is None:
            base = BaseUnits()
        
        self._x = base.x
        self._B = base.B
        self._n = base.n * 1e20
        self._kBT = base.kBT
        
    @property
    def x(self):
        return self._x
    
    @property
    def B(self):
        return self._B
    
    @property
    def n(self):
        """Unit of particle number density in 1/m^3."""
        return self._n
    
    @property
    def kBT(self):
        return self._kBT
    
    @property
    def v(self):
        """Unit of velocity in m/s."""
        return self._v
    
    @property
    def t(self):
        """Unit of time in s."""
        return self._t
    
    @property
    def p(self):
        """Unit of pressure in Pa, equal to B^2/mu0 if velocity_scale='alfvén'."""
        return self._p
    
    @property
    def rho(self):
        """Unit of mass density in kg/m^3."""
        return self._rho
    
    @property
    def j(self):
        """Unit of current density in A/m^2."""
        return self._j
    
    def derive_units(self, velocity_scale: str = "light", A_bulk: int = None, Z_bulk: int = None,
                     verbose=False):
        """Derive the remaining units from the base units, velocity scale and bulk species' A and Z."""

        from mpi4py import MPI

        con = ConstantsOfNature()

        # velocity (m/s)
        if velocity_scale is None:
            self._v = 1.0

        elif velocity_scale == "light":
            self._v = con.c

        elif velocity_scale == "alfvén":
            assert A_bulk is not None, 'Need bulk species to choose velocity scale "alfvén".'
            self._v = self.B / np.sqrt(self.n * A_bulk * con.mH * con.mu0)

        elif velocity_scale == "cyclotron":
            assert Z_bulk is not None, 'Need bulk species to choose velocity scale "cyclotron".'
            assert A_bulk is not None, 'Need bulk species to choose velocity scale "cyclotron".'
            self._v = Z_bulk * con.e * self.B / (A_bulk * con.mH) * self.x

        elif velocity_scale == "thermal":
            assert A_bulk is not None, 'Need bulk species to choose velocity scale "thermal".'
            assert self.kBT is not None
            self._v = np.sqrt(self.kBT * 1000 * con.e / (con.mH * A_bulk))

        # time (s)
        self._t = self.x / self.v
        
        # return if no bulk is present
        if A_bulk is None:
            self._p = None
            self._rho = None
            self._j = None
        else:
            # pressure (Pa), equal to B^2/mu0 if velocity_scale='alfvén'
            self._p = A_bulk * con.mH * self.n * self.v ** 2

            # mass density (kg/m^3)
            self._rho = A_bulk * con.mH * self.n

            # current density (A/m^2)
            self._j = con.e * self.n * self.v
        
        # print to screen
        if verbose and MPI.COMM_WORLD.Get_rank() == 0:
            units_used = (" m", " T", " m⁻³", "keV", " m/s", " s", " bar", " kg/m³", " A/m²",)
            print("")
            for (k, v), u in zip(self.__dict__.items(), units_used):
                if v is None:
                    print(f"Unit of {k[1:]} not specified.")
                else:
                    print(
                        f"Unit of {k[1:]}:".ljust(25),
                        "{:4.3e}".format(v) + u,
                    )


@dataclass
class DerhamOptions:
    """Options for the Derham spaces.

    Parameters
    ----------
    p : tuple[int]
        Spline degree in each direction.

    spl_kind : tuple[bool]
        Kind of spline in each direction (True=periodic, False=clamped).

    dirichlet_bc : tuple[tuple[bool]]
        Whether to apply homogeneous Dirichlet boundary conditions (at left or right boundary in each direction).
        
    nquads : tuple[int]
        Number of Gauss-Legendre quadrature points in each direction (default = p, leads to exact integration of degree 2p-1 polynomials).

    nq_pr : tuple[int]
        Number of Gauss-Legendre quadrature points in each direction for geometric projectors (default = p+1, leads to exact integration of degree 2p+1 polynomials).
    
    polar_ck : PolarRegularity
        Smoothness at a polar singularity at eta_1=0 (default -1 : standard tensor product splines, OR 1 : C1 polar splines)

    local_projectors : bool
        Whether to build the local commuting projectors based on quasi-inter-/histopolation.
    """
    p: tuple = (1, 1, 1)
    spl_kind: tuple = (True, True, True)
    dirichlet_bc: tuple = ((False, False), (False, False), (False, False))
    nquads: tuple = None
    nq_pr: tuple = None
    polar_ck: PolarRegularity = -1
    local_projectors: bool = False

    def __post_init__(self):
        check_option(self.polar_ck, PolarRegularity)


@dataclass
class FieldsBackground:
    """Options for backgrounds in configuration (=position) space.
 
    Parameters
    ----------
    type : BackgroundTypes
        Type of background.
        
    values : tuple[float]
        Values for LogicalConst on the unit cube. 
        Can be length 1 for scalar functions; must be length 3 for vector-valued functions.
        
    variable : str
        Name of the function in FluidEquilibrium that should be the background.
    """

    type: BackgroundTypes = "LogicalConst"
    values: tuple = (1.5, 0.7, 2.4)
    variable: str = None

    def __post_init__(self):
        check_option(self.type, BackgroundTypes)
        
        
@dataclass
class EnvironmentOptions:
    """Environment options for launching run on current architecture 
    (these options do not influence the simulation result). 

    Parameters
    ----------
    out_folders : str
        The directory where all sim_folders are stored. 
        
    sim_folder : str
        Folder in 'out_folders/' for the current simulation (default='sim_1').
        Will create the folder if it does not exist OR cleans the folder for new runs.
        
    restart : bool
        Whether to restart a run (default=False).

    max_runtime : int,
        Maximum run time of simulation in minutes. Will finish the time integration once this limit is reached (default=300).

    save_step : int
        When to save data output: every time step (save_step=1), every second time step (save_step=2), etc (default=1).

    sort_step: int, optional
        Sort markers in memory every N time steps (default=0, which means markers are sorted only at the start of simulation)

    num_clones: int, optional
        Number of domain clones (default=1)
    """

    out_folders: str = os.getcwd()
    sim_folder: str = "sim_1"
    restart: bool = False
    max_runtime: int = 300
    save_step: int = 1
    sort_step: int = 0
    num_clones: int = 1
    
    def __post_init__(self):
        self.path_out: str = os.path.join(self.out_folders, self.sim_folder)
        
    def __repr__(self):
        for k, v in self.__dict__.items():
            print(f"{k}:".ljust(20), v)
    
    
def check_option(opt, options):
    """Check if opt is contained in options; if opt is a list, checks for each element."""
    opts = get_args(options)
    if not isinstance(opt, list):
        opt = [opt]
    for o in opt:
        assert o in opts, f"Option '{o}' is not in {opts}." 

